sklearn/linear_model/_glm/glm.py

```diff
@@ -0,0 +1,885 @@
+"""
+Generalized Linear Models with Exponential Dispersion Family
+"""
+
+# Author: Christian Lorentzen <lorentzen.ch@googlemail.com>
+# some parts and tricks stolen from other sklearn files.
+# License: BSD 3 clause
+
+import numbers
+
+import numpy as np
+import scipy.optimize
+
+from ...base import BaseEstimator, RegressorMixin
+from ...utils import check_array, check_X_y
+from ...utils.optimize import _check_optimize_result
+from ...utils.validation import check_is_fitted, _check_sample_weight
+from .distribution import (
+        ExponentialDispersionModel,
+        TweedieDistribution,
+        EDM_DISTRIBUTIONS
+)
+from .link import (
+        Link,
+        IdentityLink,
+        LogLink,
+)
+
+
+class GeneralizedLinearRegressor(BaseEstimator, RegressorMixin):
+    """Regression via a Generalized Linear Model (GLM) with penalties.
+
+    GLMs based on a reproductive Exponential Dispersion Model (EDM) aim at
+    fitting and predicting the mean of the target y as mu=h(X*w). Therefore,
+    the fit minimizes the following objective function with L2
+    priors as regularizer::
+
+            1/(2*sum(s)) * deviance(y, h(X*w); s)
+            + 1/2 * alpha * |w|_2
+
+    with inverse link function h and s=sample_weight.
+    The parameter ``alpha`` corresponds to the lambda parameter in glmnet.
+
+    Read more in the :ref:`User Guide <Generalized_linear_regression>`.
+
+    Parameters
+    ----------
+    alpha : float, optional (default=1)
+        Constant that multiplies the penalty terms and thus determines the
+        regularization strength.
+        See the notes for the exact mathematical meaning of this
+        parameter.``alpha = 0`` is equivalent to unpenalized GLMs. In this
+        case, the design matrix X must have full column rank
+        (no collinearities).
+
+    fit_intercept : boolean, optional (default=True)
+        Specifies if a constant (a.k.a. bias or intercept) should be
+        added to the linear predictor (X*coef+intercept).
+
+    family : {'normal', 'poisson', 'gamma', 'inverse.gaussian'} \
```

Comment:
Most existing packages use the `family` parameter rather than `distribution` parameter for GLMs: [stastmodels](https://www.statsmodels.org/stable/glm.html), [H20](http://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/glm.html), [R](https://www.rdocumentation.org/packages/stats/versions/3.6.1/topics/glm), [py-glm](https://github.com/madrury/py-glm).

I think it's due to the fact, Tweedie is not a distribution but a family of distributions parameterized by  parametrized by a parameter p (https://en.wikipedia.org/wiki/Tweedie_distribution). Although `TwedieDistrubution(p=2)` is a distribution not a family.

> I'm more familiar with distribution, and I think that's what we usually use in scikit-learn.

The only other place we using it is `QuantileTransformer` with `output_distribution`. I think I would rather use the same parameter name as the other existing GLM packages.